### PR TITLE
MULE-18029: Fine grained Error resolution for the runtime core

### PR DESCRIPTION
**Pull request by IvanAndresFritzler**
This enhancement introduces a new MuleException that wraps the underlying causes as suppressed from the runtime Error resolution.

### CODE REVIEW

File: src/main/java/org/mule/runtime/api/exception/SuppressedMuleException.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  private static final long serialVersionUID = -2020531237382360468L;
+
+  /**
+   * Constructs a new {@link SuppressedMuleException}
+   * @param causeToSuppress The cause that wants to be suppressed. Cannot be null.
+   */
+  public SuppressedMuleException(Exception causeToSuppress) {
+    super(causeToSuppress);
+    if (causeToSuppress == null) {
```
Comment (elrodro83) [inline src/main/java/org/mule/runtime/api/exception/SuppressedMuleException.java]: Use `Objects.requireNotNull`

File: src/main/java/org/mule/runtime/api/exception/ExceptionHelper.java [hunk 2]
```diff
@@ -138,7 +138,7 @@ public static MuleException getRootMuleException(Throwable t) {
     // properly logged.
     Map<String, Object> muleExceptionInfo = new SmallMap<>();
 
-    while (cause != null) {
+    while (cause != null && !(cause instanceof SuppressedMuleException)) {
```
Comment (elrodro83) [inline src/main/java/org/mule/runtime/api/exception/ExceptionHelper.java]: add unit test case in this module for this logic

File: src/main/java/org/mule/runtime/api/exception/SuppressedMuleException.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.exception;
```
Comment (elrodro83) [inline src/main/java/org/mule/runtime/api/exception/SuppressedMuleException.java]: is this ok to be API? What components outside of the runtime will need access to this class?

File: src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * Wraps a provided exception as suppressed, meaning that the Mule Runtime will not take it into account for the error handling.
+ * The provided cause and all its nested {@link Exception#getCause()} will not be taken into account during the
+ * {@link org.mule.runtime.api.message.Error} resolution.
+ * <br/><br/>Example without suppression:
+ * <br/><pre>throw new {@link org.mule.runtime.api.exception.TypedException}(new {@link org.mule.runtime.api.connection.ConnectionException}(), {@link org.mule.runtime.api.message.ErrorType customErrorType})</pre>
+ * will resolve to an {@link Error} whith {@link Error#getErrorType()} returning ConnectionException's error type (discarding the top level customErrorType)
+ * <br/><br/>Same example with suppression:
+ * <br/><pre>throw new {@link org.mule.runtime.api.exception.TypedException}(new {@link SuppressedMuleException}(new {@link org.mule.runtime.api.connection.ConnectionException}()), {@link org.mule.runtime.api.message.ErrorType customErrorType})</pre>
+ * will resolve to an {@link Error} with {@link Error#getErrorType()} returning customErrorType (discarding the underlying ConnectionException's error type})
+ * @since 4.3
```
Comment (elrodro83) [inline src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java]: `1.3, 1.2.3`

Comment (elrodro83) [reply src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java]: or `1.2.4`, don't remember if we are already in code freeze

File: src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  private static final long serialVersionUID = -2020531237382360468L;
+
+  /**
+   * Constructs a new {@link SuppressedMuleException}
+   *
+   * @param causeToSuppress The cause that wants to be suppressed. Cannot be null.
+   */
+  public SuppressedMuleException(Throwable causeToSuppress) {
+    super(causeToSuppress);
+    Objects.requireNonNull(causeToSuppress, "Cannot suppress a null cause");
```
Comment (elrodro83) [inline src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java]: static import. Also, inline this in the line above

File: src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java [hunk 6]
```diff
@@ -110,6 +137,38 @@ public void registerGlobalNotFromRuntime() throws Exception {
...
+    Throwable innerCause = new ConnectionException(new NullPointerException());
+    Throwable errorWithUnsuppressedCause = new TypedException(innerCause, dummyErrorType);
+    assertThat(ExceptionHelper.getRootMuleException(errorWithUnsuppressedCause), is(innerCause));
+  }
+
+  @Test
+  public void suppressedMuleExceptionInGetRootMuleException() {
+    Throwable innerCause = new ConnectionException(new NullPointerException());
+    Throwable errorWithSuppressedCause = new TypedException(new SuppressedMuleException(innerCause), dummyErrorType);
+    assertNull(ExceptionHelper.getRootMuleException(errorWithSuppressedCause));
```
Comment (elrodro83) [inline src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java]: hamcrest

File: src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java [hunk 7]
```diff
@@ -110,6 +137,38 @@ public void registerGlobalNotFromRuntime() throws Exception {
     registerGlobalExceptionReader(cl1Reader);
   }
 
+  @Test
+  public void unsuppressedMuleExceptionInGetRootMuleException() {
+    Throwable innerCause = new ConnectionException(new NullPointerException());
+    Throwable errorWithUnsuppressedCause = new TypedException(innerCause, dummyErrorType);
+    assertThat(ExceptionHelper.getRootMuleException(errorWithUnsuppressedCause), is(innerCause));
```
Comment (elrodro83) [inline src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java]: static import

File: src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java (hunk 6 above)
Comment (elrodro83) [reply src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java]: static import

File: src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java [hunk 8]
```diff
@@ -110,6 +137,38 @@ public void registerGlobalNotFromRuntime() throws Exception {
...
+    Throwable innerCause = new ConnectionException(new NullPointerException());
+    Throwable errorWithSuppressedCause = new TypedException(new SuppressedMuleException(innerCause), dummyErrorType);
+    assertNull(ExceptionHelper.getRootMuleException(errorWithSuppressedCause));
+  }
+
+  @Test
+  public void unsuppressedMuleExceptionInGetExceptionsAsList() {
+    Throwable innerCause = new ConnectionException(new NullPointerException());
+    Throwable errorWithUnsuppressedCause = new TypedException(innerCause, dummyErrorType);
+    List<Throwable> exceptionsList = ExceptionHelper.getExceptionsAsList(errorWithUnsuppressedCause);
```
Comment (elrodro83) [inline src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java]: static import

File: src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java [hunk 9]
```diff
@@ -110,6 +137,38 @@ public void registerGlobalNotFromRuntime() throws Exception {
...
+    assertThat(exceptionsList, hasItems(errorWithUnsuppressedCause, innerCause));
+    assertThat(exceptionsList, hasSize(3));
+  }
+
+  @Test
+  public void suppressedMuleExceptionInGetExceptionsAsList() {
+    Throwable innerCause = new ConnectionException(new NullPointerException());
+    Throwable errorWithSuppressedCause = new TypedException(new SuppressedMuleException(innerCause), dummyErrorType);
+    List<Throwable> exceptionsList = ExceptionHelper.getExceptionsAsList(errorWithSuppressedCause);
```
Comment (elrodro83) [inline src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java]: static import

File: src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java [hunk 10]
```diff
@@ -110,6 +137,38 @@ public void registerGlobalNotFromRuntime() throws Exception {
...
+    assertNull(ExceptionHelper.getRootMuleException(errorWithSuppressedCause));
+  }
+
+  @Test
+  public void unsuppressedMuleExceptionInGetExceptionsAsList() {
+    Throwable innerCause = new ConnectionException(new NullPointerException());
+    Throwable errorWithUnsuppressedCause = new TypedException(innerCause, dummyErrorType);
+    List<Throwable> exceptionsList = ExceptionHelper.getExceptionsAsList(errorWithUnsuppressedCause);
+    assertThat(exceptionsList, hasItems(errorWithUnsuppressedCause, innerCause));
+    assertThat(exceptionsList, hasSize(3));
```
Comment (elrodro83) [inline src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java]: this says 3, but the previous assertion only has 2 items. Assert the other one also

Comment (elrodro83) [reply src/test/java/org/mule/runtime/api/exception/ExceptionHelperTestCase.java]: better yet, use a matcher that verifies the order of the list as well

File: src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java (hunk 4 above)
Comment (IvanAndresFritzler) [reply src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java]: 1.2.3!

File: src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java [hunk 11]
```diff
@@ -21,7 +21,7 @@
  * <br/><br/>Same example with suppression:
  * <br/><pre>throw new {@link org.mule.runtime.api.exception.TypedException}(new {@link SuppressedMuleException}(new {@link org.mule.runtime.api.connection.ConnectionException}()), {@link org.mule.runtime.api.message.ErrorType customErrorType})</pre>
  * will resolve to an {@link Error} with {@link Error#getErrorType()} returning customErrorType (discarding the underlying ConnectionException's error type})
- * @since 4.3
+ * @since 1.2.3
```
Comment (elrodro83) [inline src/main/java/org/mule/runtime/internal/exception/SuppressedMuleException.java]: 1.3 as well

**Comment by IvanAndresFritzler** [comment]
--validate

**Pull request merged**